doc/design/switch_kernel.md

```diff
@@ -0,0 +1,66 @@
+
+The `layout` is useful for some computational library. One example is that MKLDNN uses many kinds of layout, such as `nChw8`. Each kind of layout will invoke the different kernel.
+
+## Problem
+
+We register a kernel for every operator and every kernel type ideally. However, it is impracticable for the following situations.
+
+1. Some operators, like CRF, are complicated and inefficient to be implemented on GPU. The CRF operator will only have a CPU kernel.
+2. Some operators will take too many memory. It is better to force them into CPU. However, the rest of operators in this neural network will be performed in GPU, i.e., model parallel problem.
+3. Some layout and place are particular. One example is that MKLDNN uses `nChw8` and there is no other library uses `nChw8`.
+
+Problems under these situations are similar. We can formalise this problem as follow.
+
+We register kernels with types $KT = \{kt_1, kt_2, kt_3, ...\}$ for one operator. The inputs of this operator should be run on kernel type $kt_{?}$, which the $kt_{?} \notin KT$. How to cast the input of this operator from $kt_{?}$ to any of kernel type in $KT$.
+
+## Solution
+
+It is clearly that transforming inputs of an operator from a specific kernel type to another kernel type is not related to the particular operator. So we should register these transformation methods as global methods.
+
+We can infer a kernel type from the inputs of an operators. We let this kernel type as `actual kernel type`, which means this kernel type is the actually kernel type that operator should be performed.
+
+We can get a kernel type by 1) The configuration of operator description. (Users may want to force use `MKL` for `conv` operator). 2) The place of the current executor. (Executor is running on GPU). This kernel type is what we expect the operator will be performed on. We let this kernel type as `expect kernel type`.
+
+We transform the input data from `actual` to `expect` if the expect kernel type is not as same as actual kernel type.
+
+The algorithm is described as follow
+
+```cpp
+using DataTransformationFN = std::function<void(const Tensor& in, Tensor* out)>;
+using KernelTypePair = std::pair<KernelType, KernelType>;
+
+map<KernelTypePair, DataTransformationFN> g_data_transformation_;
+
+void OpWithKernel::Run() {
+  vec<Tensor> inputs = ...
+  auto actual_kernel_type = GetActualKernelType(inputs);
+  
+  // The expected kernel type is related to actual kernel type.
+  // For the most operators, the expected kernel type is as same as
+  // actual kernel type.
+  //
+  // So we pass `actual_kernel_type` as a parameter of 
+  // GetExpectedKernelType
+  auto expect_kernel_type = GetExpectedKernelType(actual_kernel_type);
+  
+  auto trans = g_data_transformation_[{actual_kernel_type, expect_kernel_type}];
+  
+  kenerl.run(trans(inputs));
```

Comment:
How about the `Tensor` after converted ?  Would it be deleted? Or should stored at somewhere in case of re-alloc?

Reply:
Well. It depends.
We only transform input data. We can re-alloc the memory and do not worry about the synchronize between the original tensor and the re-alloced tensor since it is the input of the operator and it is read-only in the operator.

We can delete the original data if the data is not used when this operator is complete.  However, I think reusing memory is another topic. We can archive that by tracing every variable's usage.

Reply:
I just want to confirm that I didn't misunderstand on this question.

 `Tensor` in different `data_transformation` function behave different.  There is 3 kind of situations.
For layout transform, `nChw8c ` => `NCHW`, we have to re-alloc a memory and assign the new Tensor with correct layout. 
For Datatype transform, `int`  => `float`, just reuse the Tensor.
For Place transform, if we support `Place Fallback`(I mean, will try to use the kernel by order <CUDNNconv, gemmGPU, gemmCPU >).  Do we allow GPUPlace fallback to CPUPlace? if so, we also need to re-alloc a CPU memory and make a copy.

Since the all the conversion will happen in every training instance, so the old-Tensor and new-Tensor must be re-used.

Reply:
> For Datatype transform, int => float, just reuse the Tensor.

**Reusing memory is another topic.**

We cannot reuse Tensor if the tensor is input data of other operators. Reusing memory should take care of the whole neural network. If the input is not used by other operators after the current operator, we can free the memory of this tensor.

We can give a quick implementation. The simplest way to implement the transformation of the tensor is always creating a new tensor.

There is no conflict between creating a new tensor while transforming and reusing memory. They are two independent topics. We can reuse memory even we create a new tensor every mini-batch.

Reply:
Thx @reyoung , indeed I think its another topic.

Looking forward to that.